doc/api/process.md

```diff
@@ -1881,6 +1881,210 @@ a code.
 Specifying a code to [`process.exit(code)`][`process.exit()`] will override any
 previous setting of `process.exitCode`.
 
+## `process.finalization.register(ref, callback)`
+
+<!-- YAML
+added: REPLACEME
+-->
+
+> Stability: 1 - Experimental
+
+* `ref` {Object | Function} The reference to the resource that is being tracked.
+* `callback` {Function} The callback function to be called when the resource
+  is finalized.
+  * `ref` {Object | Function} The reference to the resource that is being tracked.
+  * `event` {string} The event that triggered the finalization. Defaults to 'exit'.
+
+This function registers a callback to be called when the resource is
+finalized, via garbage collection or when the process emits the `exit` event.
+
+Inside the callback, you can do whatever you want with the object,
+including freeing resources allocated by that resource.
+
+Caveats: This function relies on the `FinalizationRegistry` API, which the
+authors of this API said to "avoid where possible" because the nature of
+how the garbage collector works differently between engines, there's no
+specific behavior guarantee by the specification. In their words:
+"Developers shouldn't rely on cleanup callbacks for essential program
+logic. Cleanup callbacks may be useful for reducing memory usage across
+the course of a program, but are unlikely to be useful otherwise". You
+can read more about the caveats at [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/FinalizationRegistry#avoid_where_possible)
+or in the [proposal WeakRefs](https://github.com/tc39/proposal-weakrefs?tab=readme-ov-file#a-note-of-caution).
+
+Although you don't know exactly when the callback will be called because of
+the GC, you can be sure the callback will be called at least once when the
+process exits.
+
+If the object is garbage collected before the `exit` event is emitted, the
+callback will be called and will unregister the object from the finalization
+registry to not be called again.
+
+```cjs
+const { finalization } = require('node:process');
+
+const myDisposableObject = {
+  dispose() {
+    // Free your resources synchronously
+  },
+};
+
+function onFinalize(obj, event) {
+  // You can do whatever you want with the object
+  obj.dispose();
+}
+
+finalization.register(myDisposableObject, onFinalize);
+```
+
+```mjs
+import { finalization } from 'node:process';
+
+const myDisposableObject = {
+  dispose() {
+    // Free your resources synchronously
+  },
+};
+
+function onFinalize(obj, event) {
+  // You can do whatever you want with the object
+  obj.dispose();
+}
+
+finalization.register(myDisposableObject, onFinalize);
+```
```

Comment:
This example _will not work_, because the closure is "closed" on the object. See the on-exit-leak free example.

Reply:
I didn't understand what you mean.

In the example, the object is probably unlikely to be GC, but it will call the dispose when the process exit.

Do you want an example where the object could be GC?

Reply:
The object will _never_ GC, because a reference is kept in the callback. A strong reference is kept on the callback.

The key point of this feature is to allow the object to be garbage collected, but also to be disposed of during process shutdown.

Reply:
Initially, I thought the callback was called when the object is GC, but that's not true.

I updated the docs and the examples to try to be more explicit on this behavior.